shell/browser/api/electron_api_web_frame_main.cc

```diff
@@ -0,0 +1,230 @@
+  gin_helper::Promise<base::Value> promise(args->isolate());
+  v8::Local<v8::Promise> handle = promise.GetHandle();
+
+  if (render_frame_disposed_) {
+    promise.RejectWithErrorMessage(
+        "Render frame was disposed before WebFrame.executeJavaScript could be "
+        "executed");
+    return handle;
+  }
+
+  if (has_user_gesture) {
+    auto* ftn = content::FrameTreeNode::From(render_frame_);
+    ftn->UpdateUserActivationState(
+        blink::mojom::UserActivationUpdateType::kNotifyActivation,
+        blink::mojom::UserActivationNotificationType::kTest);
+  }
+
+  render_frame_->ExecuteJavaScriptForTests(
+      code, base::BindOnce([](gin_helper::Promise<base::Value> promise,
+                              base::Value value) { promise.Resolve(value); },
+                           std::move(promise)));
+
+  return handle;
+}
+
+bool WebFrame::Reload(gin::Arguments* args) {
+  CHECK_RENDER_FRAME(reload, args, false);
+  return render_frame_->Reload();
+}
+
+int WebFrame::FrameTreeNodeID(gin::Arguments* args) const {
+  CHECK_RENDER_FRAME(frameTreeNodeId, args, -1);
+  return render_frame_->GetFrameTreeNodeId();
+}
+
+int WebFrame::ProcessID(gin::Arguments* args) const {
+  CHECK_RENDER_FRAME(processId, args, -1);
+  return render_frame_->GetProcess()->GetID();
+}
+
+int WebFrame::RoutingID(gin::Arguments* args) const {
+  CHECK_RENDER_FRAME(routingId, args, -1);
+  return render_frame_->GetRoutingID();
+}
+
+GURL WebFrame::URL(gin::Arguments* args) const {
+  CHECK_RENDER_FRAME(url, args, GURL::EmptyGURL());
+  return render_frame_->GetLastCommittedURL();
+}
+
+content::RenderFrameHost* WebFrame::Top(gin::Arguments* args) {
+  CHECK_RENDER_FRAME(top, args, nullptr);
+  return render_frame_->GetMainFrame();
+}
+
+content::RenderFrameHost* WebFrame::Parent(gin::Arguments* args) {
+  CHECK_RENDER_FRAME(parent, args, nullptr);
+  return render_frame_->GetParent();
+}
+
+std::vector<content::RenderFrameHost*> WebFrame::Frames(gin::Arguments* args) {
+  std::vector<content::RenderFrameHost*> frame_hosts;
+  CHECK_RENDER_FRAME(frames, args, frame_hosts);
+
+  for (auto* rfh : render_frame_->GetFramesInSubtree()) {
+    if (rfh->GetParent() == render_frame_)
+      frame_hosts.push_back(rfh);
+  }
+
+  return frame_hosts;
+}
+
+std::vector<content::RenderFrameHost*> WebFrame::FramesInSubtree(
+    gin::Arguments* args) {
+  std::vector<content::RenderFrameHost*> frame_hosts;
+  CHECK_RENDER_FRAME(frames, args, frame_hosts);
+
+  for (auto* rfh : render_frame_->GetFramesInSubtree()) {
+    frame_hosts.push_back(rfh);
+  }
+
+  return frame_hosts;
+}
+
+// static
+gin::Handle<WebFrame> WebFrame::From(v8::Isolate* isolate,
+                                     content::RenderFrameHost* rfh) {
+  if (rfh == nullptr)
+    return gin::Handle<WebFrame>();
+  auto frame_map = g_render_frame_map.Get();
+  auto iter = frame_map.find(rfh);
+  WebFrame* web_frame =
+      iter == frame_map.end() ? new WebFrame(rfh) : iter->second;
```

Comment:
Perhaps refactor this lookup into a common function? The code is repeated in a few places and I think a `WebFrame* FromRenderFrameHost(content::RenderFrameHost* rfh)` function would aid readability.